### ISSUE TITLE
Finding global variables (GC implementation)

### ISSUE THREAD

**Issue by Maaarcocr** [zetavm/zetavm#31]
In the process of trying to make a very simple GC, or at least starting trying out things for it, I wanted to print all the declared variables when VM::alloc is called. 

I've been successful to print variables that are in the stack with this function: 
```c++
void getStackVariables() {
    for (Value* ptr = stackPtr; ptr != stackBase; ptr++) 
    {
        if (ptr[0].isPointer())
        {
            std::cout << ptr[0].toString() << std::endl;
            if (ptr[0].isObject())
                {
                for (auto itr = ObjFieldItr(ptr[0]); itr.valid(); itr.next())
                {
                    auto fieldName = itr.get();
                    std::cout << "field: " << fieldName << std::endl;
                }
            }
            else if (ptr[0].isString()) 
            {
                std::cout << "string: " << (std::string)ptr[0] << std::endl; 
            }
        }
    }
}
```

This function is inside interp.cpp

What I'm doing is just traversing the stack and printing the Values that are pointers (the Values I'm interested in  garbage collecting). This works fine, but I can't find global variables when this function is called when the interpreter is working on a function.

I understand that global variables in Plush are fields of the @global_object, but I can't seem to find where this object is store during execution. What I want to do is finding all global variables that are active when calling the function I've written and check if they are pointers. If they are then it'll print them.

I'm sorry if the description happens not to be clear :/

**Comment by werediver** [comment]
After a quick look it seems that the VM does not track heap allocations at the moment (I would expect to see some tracking in `VM::alloc()`). Perhaps, it's a feature to be implemented as a part of GC development.

**Comment by Maaarcocr** [comment]
@werediver Yes, I'm aware of that. But I don't think that global variable are heap allocated, such as Int32 global variables. The global object has to be somewhere during execution, if not it would not be possible to access its values. 

I hope that now the issue is a bit clearer :)

**Comment by maximecb** [comment]
References to objects that you seed a garbage collector algorithm with are called roots in GC parlance.

There isn't just one global object. Each package/module has its own. The global object is directly referenced by push instructions, and pushed on the stack when it needs to be accessed. I was thinking I might want to add a pointer to the current function to function arguments, in order to facilitate the implementation of closures. This would solve your current predicament, because if some function can access the global object, then it must contain a reference to the global object in its code somewhere.

Sidenote: this isn't a problem for now, but things will get a bit more complicated when we implement a JIT compiler, because some values are going to be implicit, or encoded in machine code, and the stack might be left in an inconsistent state. We will need to implement a mechanism to safely access live variables.

@werediver That's right, there is no GC implementation right now. Memory is malloc'ed until things explode, or the process terminates, whichever happens first.

In terms of GC, I would like to try a non-moving GC, because this will allow us to "bake" pointer in machine code without needing to update them later, and it will offer better performance when processes are forked. A copying collector works poorly with forking because the OS will typically copy pages when they are written to, and a copying GC will rewrite the entire heap.

**Comment by Maaarcocr** [comment]
I see.

So when the global object get pushed, how does C++ find it? Like, the interpreter must have some sort of reference to it in order to be able to push it to the stack, or am I misunderstanding something?

**Comment by maximecb** [comment]
The reference is in the object that represents the push instruction.

{ op:'push', val:@global_obj} // <== this instruction is an object, val is a property that refers to global_obj

**Comment by Maaarcocr** [comment]
There is still something that it's not clear to me.

So when @global_obj get pushed, where is its value taken from? Like @global_obj is just a string, where does the Value, that is pushed to the stack, come from? Like the reference to it. 

I'm sorry to ask so much about this, but I really can't wrap my head around it.

**Comment by maximecb** [comment]
The zim files describe a graph of objects and values. When the files are loaded, the graph gets constructed. Once in memory, there are no strings, the instruction directly contains a reference to the global object. See `vm/parser.cpp` for more details on how that works.

**Comment by Maaarcocr** [comment]
Oh okay, I'll have a look! Thanks!

**Comment by Maaarcocr** [comment]
I had a look and now it's cleared how it works. 
Right now it seems like accessing global definitions and using them as roots for the GC would be quite difficult, as they are directly substituted during parsing. 

Even with what you propose, a function having references to what it uses, a safe GC may not be feasible. For example a function may not use a global variables, but the next one will. (I may have misunderstood your proposal) 

```
var x = {a: 10};
var f1 = function () {...}; //doesn't use x
var f2 = function () {...}; //uses x
f1();
f2();
```

I was thinking that maybe when this global object are created (in the parser) I could set a flag (that I would add) in their Value object that sets them as global. Globals should never be garbage collected, right?

**Comment by maximecb** [comment]
You need to understand the concept of "liveness". Data is live if it can be reached from the set of "root" pointers we provide to the GC traversal/marking algorithm. Liveness is also transitive.

Right now, the global object is referenced inside the bytecode of functions. Any function that can access the global object will need to get a reference to it somehow. Either through its own bytecode, with a push instruction placing it on the stack when needed, or because the global object was passed as an argument.

If I change the stack so that every function that is currently on the stack gets an argument that is a reference to itself (a "closure" argument), then the global object will indirectly be reachable through the code of those functions (stack refers to functions, functions refer to global object, global object contains global variables). If no function has a reference to the global object, then no function can access it, and that means the global object is dead/garbage, and ready to be collected.

I will look into adding a hidden function/closure argument to function calls. This will solve the problem you are currently faced with.

**Comment by Maaarcocr** [comment]
I see. 
I thought that global variable was love throughout the entire program, till its end.

Thanks you for explaining! :)

**Comment by maximecb** [comment]
@Maaarcocr I've been pretty busy with work and haven't gotten around to fixing that issue. May be fairly busy in the next few day as I'll be moving into a new apartment.

In the meantime, maybe this can keep you busy: https://github.com/maximecb/zetavm/issues/32

**Comment by Maaarcocr** [comment]
@maximecb I've actually played around with some basic blocks for the GC (such as making the VM keep track of everything that it allocates memory for). 

I'll work on floats in plush (as I've already mentioned in the other issue) before going back to the GC.

**Comment by Maaarcocr** [comment]
A small update on this: I've been quite busy this week, I'm still planning to implement the GC :) 

Also, from next week I'll start my summer internship (🎉), so I'll probably have less time :( but I'd still love to contribute in my free time!

**Comment by maximecb** [comment]
Congrats on your summer internship. I hope you still manage to find time because you've been extremely helpful. I most likely still wouldn't have gotten around to adding floating-point support. I'm also pretty busy with my new job. I try to do a little bit every night, even if it's just 15 minutes. Over the last couple of days, I completed support for exceptions in the VM and in cplush. That will allow me to have failing plush parser tests, and to address the problem you reported where you currently can't have a variable named assertFP, or returnFoo.

Once I'm done fixing the parser problems, I will probably add the hidden closure argument I had mentioned before, which will be useful for the GC, and for implementing closures. I'm thinking I can add closures to Plush as an optional language extension, and I have some ideas for a functional audio API that will make use of them.

I know you just said you were likely going to have less free time soon, but I've been meaning to ask you, would you be interested in implementing a minimalist lisp-like language, or forth, or any other language for zeta? I'd like to make the case for zeta as a multi-language VM, but right now there is only Plush. People contributed beginnings of other language implementations, but I'm afraid they will never finish what they started. You'd have carte blanche to implement any language you fancy, be it an existing one or your own invention.

**Comment by Maaarcocr** [comment]
I would love to implement a language. I may switch to that while you work on the closers argument :) as I said next week I'll start working and it's really my first time in industry, so I need to "adapt" my schedule. I will work on the langauge! 

I won't abandon zetaVM, so far it has been probably the best project I've worked on, so it would be stupid! 

Please, let me also know when you'll need help with the faster interpreter (by opening an issue or whatever), I would also love to help there :)

**Comment by maximecb** [comment]
Now that we have a hidden closure argument implemented (see #36), it should be possible to access all functions currently live by traversing the stack.

**Comment by maximecb** [comment]
Closing this for now, but feel free to open other GC issues when you resume work on that task (or to reopen this one).
